src/Discovery/Walk.hs

```diff
@@ -124,6 +125,10 @@ fileName = toFilePath . filename
 findFileNamed :: String -> [Path a File] -> Maybe (Path a File)
 findFileNamed name = find (\f -> fileName f == name)
 
+-- |Find the first file that matches any of the provided names.
+findFirstMatchingFile :: [String] -> [Path a File] -> Maybe (Path a File)
+findFirstMatchingFile names = find (\f -> fileName f `elem` names)
```

Comment:
Given `findFirstMatch :: [a] -> [b] -> Maybe b`
Searching all `b`s for one matching any `a` has to rescan all of `[a]` for potentially every `b` (unless GHC is way smarter than I think it is- if you think this is the case please let me know!)

Please convert `names` to a set for faster matches:
```suggestion
-- | Find the first file that matches any of the provided names.
findFirstMatchingFile :: [String] -> [Path a File] -> Maybe (Path a File)
findFirstMatchingFile names = find (\f -> fileName f `Set.member` names')
  where
    names' = Set.fromList names
```

Reply:
GHC won't do something smart to my knowlege, my reasoning here was that I wasn't using this function for more than two filenames so the `Set` wouldn't be much benefit. Using a `Set` is much better from an API perspective though since who knows what future folks will do with it.

Reply:
Yeah, I agree with your intuition for sufficiently small `a`s, but since this is a general function better to be safe.
Either that, or just throw a comment in there calling out this as a potential performance footgun.

Reply:
I committed your Set change, after you asked for it I understood that it makes more sense than what I was doing since it's in such a widely used module. Thanks!